finite_approx/gibbs_sampler_lib.py

```diff
@@ -101,3 +103,213 @@ def sample(self, burnin, num_gibbs_draws):
+
+                # flip z[n,k]
+                self.z[n,k] = 1 - self.z[n,k]
+                [log_lh_X_update, var_inv_update] = update_x_lp_cond_z(\
+                    self.x, self.x2, self.z2, self.z, self.var_inv, self.sigma_eps, \
+                    self.sigma_a, self.k_approx, n, k)
+
+                log_p_znk_cond_z_update = np.log(p_znk_cond_z_update)
+                log_p_zn_update = log_lh_X_update + log_p_znk_cond_z_update \
+                    - sp.misc.logsumexp([log_lh_X_update + log_p_znk_cond_z_update,\
+                        self.log_lh_X+ np.log(p_znk_cond_z_stay)])
+
+                #assert (np.exp(log_p_zn_update) >= 0) & (np.exp(log_p_zn_update) <= 1)
+
+                choice = np.random.binomial(1, np.exp(log_p_zn_update))
+
+                # update Z and prep log likelihood for next round:
+                if choice == 1: # update the necessary quantities
+                    self.log_lh_X = deepcopy(log_lh_X_update)
+                    self.var_inv = deepcopy(var_inv_update)
+                    self.z2 = np.dot(self.z.T, self.z)
+                else:
+                    # flip back to the orginal value of z[n,k]
+                    self.z[n,k] = 1 - self.z[n,k]
+
+        if keep_draw:
+            self.z_draws.append(self.z)
+
+    def sample(self, burnin, num_gibbs_draws):
+
+        print('Sampling:')
+        bar = progressbar.ProgressBar(max_value=num_gibbs_draws + burnin)
+        for n in bar(range(num_gibbs_draws + burnin)):
+            self.draw_z(keep_draw = n >= burnin)
+
+        print('Done sampling :)')
+
+
+def update_x_lp_cond_z(x, x2, z2, z_update, var_inv, sigma_eps, sigma_a, \
+                        K_approx, n,k):
+# likelihood p(X|Z)-- equation (8) in Griffiths and Ghahramani
+# http://mlg.eng.cam.ac.uk/zoubin/papers/ibp-nips05.pdf
+# outputs the likelihood at Z  when Z has component n,k flipped
+# var_inv refers to inv(Z^T * Z + sigma_eps/sigma_a I), the previous inverse
+
+    x_d = np.shape(x)[1]
+    x_n = np.shape(x)[0]
+    k_approx = np.shape(z_update)[1]
+
+    assert np.shape(x)[0] == np.shape(z_update)[0]
+
+    inv_var_flip = update_inv_var(z2, z_update, var_inv, sigma_eps, sigma_a, n, k)
+
+    mean_a = np.dot(np.dot(inv_var_flip, z_update.T), x)
+
+    log_likelihood = -1/(2*sigma_eps) * \
+            np.trace(x2 - np.dot(x.T, np.dot(z_update, mean_a)) )
+
+    [_, logconst] = np.linalg.slogdet(inv_var_flip)
+
+    return log_likelihood - (-x_d/2) * logconst, inv_var_flip
+
+def update_inv_var(z2, z_update, var_inv, sigma_eps, sigma_a, n, k):
+    # compute the inverse of Z^T * Z + sigma_eps/sigma_a * I when Z has
+    # element n,k flipped.
+    # var_inv refers to the previous computation, before Z had its
+    # (n,k) component flipped
+    # z2 refers to Z^T * Z
+
+    k_approx = np.shape(z2)[0]
+
+    var = z2 + sigma_eps/sigma_a * np.eye(k_approx)
+    var_flip = np.dot(z_update.T, z_update) + sigma_eps/sigma_a * np.eye(k_approx)
```

Comment:
Same here, only one term of this dot product changed since last time you called it.  Most of these operations are wasted.